### PR TITLE
Migration to Python 3 only draft

### PR DESCRIPTION
**Pull request by Carreau**
# TL;DR:

This are rough estimate timelines. 
- ASAP: Fix Pip to refuse to install Python 3 only packages on Python 2. Extremely
  high priority as we want pip version to be release and propagate after
  that.
- ASAP: Fix Setuptools to be able to express `requires-python` in `setup.py`.
- June 2016 : Release IPython 5.0 LTS
  - Traitlets "LTS"
  - ipykernel LTS
  - ipython_genutils LTS
- January 2017 : Release IPython 6.0 Python 3 only. 
- July 2017 : Release Notebook 6.0 Python 3 only.
- December 2017: IPython 7.0 
- December 2017: End of active support IPython 5.x LTS
- July 2019 : Complete end of support for IPython 5.x

### CODE REVIEW

**Comment by takluyver** [comment]
> ASAP: Fix Setuptools to be able to express requires-python in setup.py.

I don't think this is worth spending time on, given that:
- ~no-one knows about that metadata field
- we couldn't start using it until a new version of setuptools had propagated
- we can express similar information in classifiers, which are already widely used

File: migration-to-python-3-only.md [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+    ipykernel only if it dirrectly affect IPython
+
+  - pass 36 month after IPython 5.0 release, IPython 5.x will be considers
+    unsupported.
+
+  - The above time lines are intent, and the IPython core team reserve the
+    right to modify them in the future.
+
+  - regardless of above statement, Python 2 support in the 5.x branch will not
+    exceed support of CPython 2.7, and be likely much shorter.
```
Comment (takluyver) [inline migration-to-python-3-only.md]: It's not likely to be _much_ shorter - the timeline you proposed only has us dropping support for 2.7 a year before Python itself does.

File: migration-to-python-3-only.md [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    also accept to provide reasonable support requests for the IPython 5.x
+    branch.
+
+
+  - For the 36 month following the release of IPython 5.x, the core team will
+    continue to review critical & security releated patches, as well as
+    improvment to the documentation, and make related release if necessary. The
+    core team will make sure that the 5.x Version of IPython stay visible
+    various web pages mentionning how to get IPython. The core team reserve the
+    right to ignore and or close any support request, request for patch
```
Comment (takluyver) [inline migration-to-python-3-only.md]: This statement feels a bit too formal to me. We're writing a statement of intent, not a legally binding contract.

File: migration-to-python-3-only.md [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+the stable version of these software is supported, without guaranties of timeline.
+
+
+### Source distribution installation.
+
+
+The Python package manager (`pip`) does not respect the `requires-python`
+metadata field, and `setuptools` does not provide a way to set a value of this
+metadata field as far as we know. This implies that if the IPython core team release a version of
+IPython that is not compatible with IPython, the end use installing ipython
```
Comment (takluyver) [inline migration-to-python-3-only.md]: I'm guessing that one of the IPythons here should be 'Python 2'. We're doing something wrong if IPython ends up incompatible with IPython ;-)

File: migration-to-python-3-only.md (hunk 1 above)
Comment (jankatins) [inline migration-to-python-3-only.md]: With this statement it becomes quite a bit more likely [edit] that a shorter lifespan is much more acceptable because other projects will do the same and so users will do it too[/]:-)

File: migration-to-python-3-only.md [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+inconsistency for notebook server extensions, as on Python 2 Anaconda, the
+notebook server will be in a separate environment, and the notebook extensions
+and optional dependencies need to be installed in this environment as well. We
+are thus speaking of `nbconvert`, custom exporters, custom handler and server
+side extension.
+
+The anaconda crew told us that dropping support for Python 2 will lead them to
+unbundle Jupyter from Anaconda instead of using the 2 environment trick.
+
+This is not a problem on linux and unix-like system where notebook will "just" requires Python 3.
```
Comment (takluyver) [inline migration-to-python-3-only.md]: This isn't very clear: it's not the unixy-ness that matters here, it's how different packaging systems have chosen to handle Python 2 and 3. It's still a problem for Anaconda on Unixy systems, but it's not a problem for apt/rpm, which treat Python 2 and 3 as two different things.

(Of course, it's anyone's guess when, if ever, Jupyter will be packaged in Debian. But this is one thing that won't be a problem for them.)

**Comment by jankatins** [comment]
Thanks for starting "the great migration" :-)

File: migration-to-python-3-only.md (hunk 1 above)
Comment (Carreau) [inline migration-to-python-3-only.md]: A year before is 25%  :-P

**Comment by Carreau** [comment]
Thanks @takluyver for the ~~few fixes~~ rewriting everything :-P

**Comment by Carreau** [comment]
cc @fperez

File: migration-to-python-3-only.md [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+
+
+## Challenges
+
+As for IPython above; the notebook package need to be able to express that it
+requires Python 3. This is mitigated by people installing `jupyter`: we can
+make that conditionally depend on different version of `notebook`, depending on
+the version of Python.
+
+In particular for Anaconda, one path forward would be to bundle the
```
Comment (damianavila) [inline migration-to-python-3-only.md]: @Carreau @takluyver is there any record of the conversation with the Anaconda team? I would like to take a look if possible. Thanks!

**Comment by damianavila** [comment]
Detailed and nice doc! :+1:

File: migration-to-python-3-only.md (hunk 5 above)
Comment (Carreau) [inline migration-to-python-3-only.md]: > @Carreau @takluyver is there any record of the conversation with the Anaconda team? I would like to take a look if possible. Thanks!

No, no public record.

**Comment by Carreau** [comment]
> I don't think this is worth spending time on, given that:
> 
> ~no-one knows about that metadata field
> we couldn't start using it until a new version of setuptools had propagated
> we can express similar information in classifiers, which are already widely used

Discussed with Nathaniel, that **think** pip is most likely to accept a PR that works with the metadata field than the classifiers, and suggest to bring that up on distutils-sig.

**Comment by Carreau** [comment]
To update: 

You can generate sdist that end in `-py3.3.tar.gz`, `-py3.4.tar.gz`, `-py3.5.tar.gz` ... that will **only** install on corresponding python version.

**Comment by Carreau** [comment]
See https://github.com/pypa/pip/issues/3182

@tacaswell I know you are a defensor of Py3, this thread might be of interest to you.

**Comment by Carreau** [comment]
Related issue in scikit-bio,  : https://github.com/biocore/scikit-bio/issues/1316

**Comment by Carreau** [comment]
Updated proposal for python3 only `sdist`s.

@asmeurer You might have some input ?

**Comment by tacaswell** [comment]
I personally like the proposed timeline and the LTS, I have been thinking along the same lines recently (and not gotten around to writing them up).

I would like to stress the importance of getting buy in from as many other projects as possible on the exact time line.  We need to all take this jump together.

**Comment by takluyver** [comment]
Agreed, we definitely want to get more projects on board. Do we know who to
contact at other projects?

File: migration-to-python-3-only.md (hunk 5 above)
Comment (asmeurer) [inline migration-to-python-3-only.md]: The `notebook` conda packages depend on Python (and every conda package that depends on Python by default depends on the major Python version). Here is `conda search notebook`:

```
notebook                     4.0.1                    py27_0  defaults
                             4.0.1                    py34_0  defaults
                             4.0.2                    py27_0  defaults
                          .  4.0.2                    py34_0  defaults
                             4.0.4                    py27_0  defaults
                          .  4.0.4                    py34_0  defaults
                             4.0.4                    py35_0  defaults
                             4.0.5                    py27_0  defaults
                             4.0.5                    py34_0  defaults
                          .  4.0.5                    py35_0  defaults
                             4.0.6                    py27_0  defaults
                             4.0.6                    py34_0  defaults
                          .  4.0.6                    py35_0  defaults
                             4.1.0                    py27_0  defaults
                             4.1.0                    py34_0  defaults
                          *  4.1.0                    py35_0  defaults
                             4.1.0                    py27_1  defaults
                             4.1.0                    py34_1  defaults
                             4.1.0                    py35_1  defaults
                             4.1.0                    py27_2  defaults
                             4.1.0                    py34_2  defaults
                             4.1.0                    py35_2  defaults
```

So this is a non-issue (and the suggestion sounds like a bad idea anyway). 

Also, package maintainers and package managers should be responsible for making packages have correct version restrictions in their dependency metadata. This isn't your problem.

File: migration-to-python-3-only.md [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+- `nbgrader`
+- `jupyterhub` is already Py3 only
+- `jupyter_console`
+
+
+# Misc
+
+We propose to move forward with signing the "no more legacy python pledge"
+
+https://gist.github.com/brettcannon/2f6926dc6a7874478ccd
```
Comment (asmeurer) [inline migration-to-python-3-only.md]: If you want to do this I would suggest taking a bit more time and input from the broader community on writing the pledge (not that there's anything wrong with this, but I get the impression that @brettcannon just cobbled that together, so it should just be considered a first draft).

File: migration-to-python-3-only.md [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+  - July 2019 : Complete end of support for IPython 5.x
+
+
+
+# Statement of the problem
+
+We recognize that many users of IPython and Jupyter still need to use a Python 2
+environment, and will still need to do so for quite some time. However, as an open
+source project, there is an overhead in maintaining Python 2 and 3 compatibility,
+and this slows progress on our goals.
```
Comment (asmeurer) [inline migration-to-python-3-only.md]: This doesn't really explain why you want to do this. But perhaps that would be better left to a blog post from the developers.

**Comment by tacaswell** [comment]
And probably buy in from the major vendors (enthought, continuum, RH (is scientific python big enough for them to notice?), debian/ubuntu (ditto RH comment)).

**Comment by rgbkrk** [comment]
> RH (is scientific python big enough for them to notice?)

IPython the console is used by many in non-scientific contexts, especially developers.

File: migration-to-python-3-only.md (hunk 6 above)
Comment (brettcannon) [inline migration-to-python-3-only.md]: Yes, it was definitely cobbled together in the middle of a Twitter conversation w/ @asmeurer and @tacaswell . :smile: 

That being said, if you all want to make it a formal thing I fully support you, but I don't know if I should be driving it as it might be viewed as biased coming from a Python core dev whose name is on the PEP listing the overall feature set of Python 3. ;)

Comment (Carreau) [inline migration-to-python-3-only.md]: > If you want to do this I would suggest taking a bit more time and input from the broader community on writing the pledge (not that there's anything wrong with this, but I get the impression that @brettcannon just cobbled that together, so it should just be considered a first draft).

Yes, I just want to make the document a bit clearer on intention and vocabulary, before pinging some Python 2 supporters on that and then having broader input from the internet.

> That being said, if you all want to make it a formal thing I fully support you, but I don't' know if I should be driving it as it might be viewed as biased coming from a Python core dev whose name is on the PEP listing the overall feature set of Python 3. ;)

I'll make it clearer that we want to likely sign something "similar to the pledge",and of course get broader feedback on it as well.

File: migration-to-python-3-only.md (hunk 7 above)
Comment (Carreau) [inline migration-to-python-3-only.md]: Good point. We'll work on that.

**Comment by Carreau** [comment]
> And probably buy in from the major vendors (enthought, continuum, RH (is scientific python big enough for them to notice?), debian/ubuntu (ditto RH comment)).

We'll have to use our network to reach into these companies. I'l see if we can get Nick Coghlan input on what would be RH position on that.

> IPython the console is used by many in non-scientific contexts, especially developers.

My **guess** is that many don't care, as long as `pip install ipython --upgrade` does not break.
They won't see the difference as ~90% of feature will be similar between IPython 5.x and 6.x.
so you will **just** get an old-stable version.

**Comment by brettcannon** [comment]
I think @Carreau is basically right: some won't care, others will be disappointed they won't get any more new features but will live with it, and yet others will scream until they turn blue and then pass out due to the lack of oxygen (at which point you can ignore them since they will have stopped yelling at that point :wink: ). The key point here is that you're not completely leaving Python 2.7 users behind, you're just not giving them new features. Continuing to support them with bugfixes allows you all to have the moral high-ground in saying "we're doing the work and supporting new features on Python 2.7 is too much of a burden, so we're moving on without leaving you totally in the lurch suddenly".

**Comment by minrk** [comment]
Good description. Keeping py2 working is something we can't reasonably stop for quite some time, but at some point giving up new features of one thing (Python) means giving up new features of another (IPython/Jupyter).

**Comment by Carreau** [comment]
I moved the document into a "drafts" folder, is there any objection to merge, leaving in the draft folder for the time being, and keep collecting feedback from people.

Thus it will be easier to actually make more changes and refinement to it with various PRs.

Then once it's considered ready and accepted, we can move it out of the draft folder.

Any Objections ?

If so, once merge, I'll contact more people (especially from the Python 2 side of the wall) to also get a bit of feedback, and not get a angry mob once we release that.

I'm also working on a not-roadmap document that explain more in detail the why Python 3,

**Comment by takluyver** [comment]
Just pushed some changes to polish it a bit. I think it's looking OK now.

When contacting people who will still want Python 2 support, let's make it very clear that the IPython LTS means Python 2 will be supported for some time - they just won't get new features (like they don't get new Python language features). I am happy to help you draft messages if you want.

**Comment by ccordoba12** [comment]
I just find out this thread, very interesting discussion. I think in Continuum we will need to support Python 2.7 until our customers require it (but I'm not speaking officially). So maybe we could agree on extending the lifetime of IPython/Notebook LTS with bugfixes for as long as Python 2 is alive? It could be an enterprise effort, just as maintenance is now for Python 2.7 :-)

Pinging @msarahan about this.

File: drafts/migration-to-python-3-only.md [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+
+This document discuss some of the various possibilities to migrate Jupyter, and
+its reference Kernel implementation IPython, toward a Python 3 only code base,
+without breaking installation for Python 2 users.
+
+The document is not meant to debate whether the Jupyter & IPython project should migrate
+to a Python 3 only code base, but the timeline of such a migration.
+
+
+# TL;DR:
```
Comment (ccordoba12) [inline drafts/migration-to-python-3-only.md]: Could you change this header to something more readable? I always forget what it means :-)

Comment (Carreau) [inline drafts/migration-to-python-3-only.md]: Done.

**Comment by Carreau** [comment]
> I just find out this thread, very interesting discussion. I think in Continuum we will need to support Python 2.7 until our customers require it (but I'm not speaking officially). So maybe we could agree on extending the lifetime of IPython/Notebook LTS with bugfixes for as long as Python 2 is alive? It could be an enterprise effort, just as maintenance is now for Python 2.7 :-)

This is not publicised, nor finished either. Please don't broadcast it to too many people, and yes we are going to get feedback from people that use Python 2 also.

**Comment by takluyver** [comment]
We're thinking of an IPython LTS with support for Python 2.7 lasting roughly as long as upstream 2.7 support does. For the notebook, we're looking at dropping support a bit quicker, because it's perfectly possible to work with a Python 2 kernel while the notebook server is running on Python 3.

See conda-forge/staged-recipes#339 for discussion of a couple of ways that might work with conda environments:
- A way to have both versions of Python together in a single environment
- A concept of 'subenvironments' whereby you can express a dependency on a command line tool to be installed in a separate conda env.

**Comment by Carreau** [comment]
Ok, I'm going to merge that (still in the draft folder) and contact a few people to get feedback.
@takluyver I'll contact you off GitHub for a list of people to contact.
